### ISSUE TITLE
Inconsistent loss when resume training with vocab size that is not divisible by world size.

### ISSUE THREAD

**Issue by weixuansun** [pytorch/torchtitan#1136]
### Bug description

When I use an tokenizer that has vocabulary size that is not divisible by parallel (or world) size, the training loss will become inconsistent after resuming.

### Versions

can be reproduced using torch2.6

Reproduce:
Use any tokenizer that has a vocabulary size that is not divisible by parallel (or world) size.
train from step 0 to step 20:

<img width="1368" alt="Image" src="https://github.com/user-attachments/assets/3e70be47-e15f-469f-b8b6-fbeb3ea6eab6" />
load step 10 checkpoint and resume training:

<img width="1421" alt="Image" src="https://github.com/user-attachments/assets/70b595f1-a034-4e29-816d-4fa9565f1648" />
As shown, step 11 and following steps have inconsistent loss.

**Comment by tianyu-l** [comment]
could you try the latest pytorch nightly? might have been fixed by https://github.com/pytorch/pytorch/pull/150490

**Comment by unlimblue** [comment]
> could you try the latest pytorch nightly? might have been fixed by [pytorch/pytorch#150490](https://github.com/pytorch/pytorch/pull/150490)

https://github.com/pytorch/pytorch/tree/c729f7dbee3be20f04a6aeac41ae0cd5be23403d

Using the latest main branch source code from April 22nd to compile PyTorch still results in the same issue.

**Comment by fegin** [comment]
This looks like a checkpointing problem. Can you confirm 1) does this only happen after checkpoint resume? 2) does this only happen when vocab size is not divisible by world size? From your description, the answers to both questions seem to be yes, but I do want to confirm. Thanks!

**Comment by weixuansun** [comment]
> This looks like a checkpointing problem. Can you confirm 1) does this only happen after checkpoint resume? 2) does this only happen when vocab size is not divisible by world size? From your description, the answers to both questions seem to be yes, but I do want to confirm. Thanks!

Yes, answers to both questions are Yes. 
If I pad the vocab size to be divisible, the loss would be consistent.

**Comment by speed1313** [comment]
I also encountered a loss discrepancy when resuming from checkpoint after setting the vocab size to be indivisible by the world size when I use FSDP.
Since the loss quickly returned to the loss before resuming, I suspect there's a bug in how the optimizer state is saved.

<img width="475" height="383" alt="Image" src="https://github.com/user-attachments/assets/55dc6e7f-46e4-4993-925a-bcb1768f421a" />

By padding the vocabulary size to be divisible by the world size, the issue was resolved.

```python
def pad_vocab_to_divisible(tokenizer, world_size: int):
    current_vocab_size = len(tokenizer)
    remainder = current_vocab_size % world_size
    if remainder == 0:
        return
    
    num_to_add = world_size - remainder 
    current_reserved_special_token_num = sum(
        1 for token in tokenizer.get_vocab().keys() if "reserved_special_token" in token
    )
    new_tokens = [
        f"<|reserved_special_token_{current_reserved_special_token_num + i}|>"
        for i in range(num_to_add)
    ]
    tokenizer.add_special_tokens({"additional_special_tokens": new_tokens})
```

**Comment by tianyu-l** [comment]
This sounds concerning, let's prioritize fixing this @wwwjn

**Comment by wwwjn** [comment]
Hey @speed1313 want to follow up on this issue, Can you provide more instructions on reproducing this issue?

I tried to reproduce the issue using lasted main branch, but the results looks good to me (FSDP degree = 4). I'm guessing it might related to this change #1424 

<img width="1870" height="903" alt="Image" src="https://github.com/user-attachments/assets/ddec174a-ec3f-4645-a1b4-c04042a0849e" />

After resume from checkpoint:
<img width="1870" height="753" alt="Image" src="https://github.com/user-attachments/assets/e3a908a6-9e37-46fa-add8-f88632dee2a7" />

**Comment by speed1313** [comment]
Hi @wwwjn, thank you for your support!
I’ve prepared a reproduction script based on the [latest main branch](https://github.com/pytorch/torchtitan/commit/a44dff1a41f6c0d8e504919ce4b1b50d05102f01), along with some instructions. Here is the [code](https://github.com/speed1313/torchtitan/commit/08fb43479eedc5016383cd4db628b9d38465a25d#diff-78cc79291e219fdfd73f7b1c7b5d442d1346f821b8add32e0e02d62597fe0ee5).
I ran it with FSDP (degree = 8), and confirmed that the bug still exists.
<img width="1307" height="702" alt="Image" src="https://github.com/user-attachments/assets/779e43b4-5227-4a2a-ab16-a050c9c367df" />

**Comment by wwwjn** [comment]
> Hi [@wwwjn](https://github.com/wwwjn), thank you for your support! I’ve prepared a reproduction script based on the [latest main branch](https://github.com/pytorch/torchtitan/commit/a44dff1a41f6c0d8e504919ce4b1b50d05102f01), along with some instructions. Here is the [code](https://github.com/speed1313/torchtitan/commit/08fb43479eedc5016383cd4db628b9d38465a25d#diff-78cc79291e219fdfd73f7b1c7b5d442d1346f821b8add32e0e02d62597fe0ee5). I ran it with FSDP (degree = 8), and confirmed that the bug still exists. <img alt="Image" width="1307" height="702" src="https://private-user-images.githubusercontent.com/69028964/471775218-779e43b4-5227-4a2a-ab16-a050c9c367df.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM3NTYzOTUsIm5iZiI6MTc1Mzc1NjA5NSwicGF0aCI6Ii82OTAyODk2NC80NzE3NzUyMTgtNzc5ZTQzYjQtNTIyNy00YTJhLWFiMTYtYTA1MGM5YzM2N2RmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI5VDAyMjgxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUwYmE3ZTNiNzdiYmZlODI3NWY0MDY5NmIzYWMyNDU4OWMyMzg2NmNkMjg0YWViMzg2MjAzM2MwZTdmZjNkNTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vFU9yv6FjVoRs9QBdf7iQdhW_ucLOxPDJ_OG1selLFM">

Thank you so much for the detailed instruction! Working on it ;)

**Comment by mingjihantencent** [comment]
Hi Torchtitan team,

I encountered the same inconsistent loss issue when training [Hunyuan-1.8B](https://huggingface.co/tencent/Hunyuan-1.8B-Instruct/tree/main) using Torchtitan FSDP2 on 8 GPUs.

In the original model, the [vocab_size](https://huggingface.co/tencent/Hunyuan-1.8B-Instruct/tree/main) is 120,818, which is not divisible by 8. When I modified the vocab size to 120,824 (divisible by 8), the inconsistent loss problem disappeared.

This suggests the issue might be related to uneven DTensor sharding. I hope this observation helps with diagnosing the problem.

Best regards,
Mingji

**Comment by wwwjn** [comment]
Hi folks, here's more update and observations:
## Hypothesis:
- Tok_embedding weight uneven sharding
- Optimizer status is not saved correctly
- FSDPModulel wrap problem


## Code: 
https://github.com/wwwjn/torchtitan/pull/new/debug_vocab_size
- Use llama3 debug model
- With seed = 0 and deterministic for reproducibility
- Total step = 20, save checkpoint at step=10. In another run, load checkpoint from step=10


## FSDP=4 (FSDP only)
<img width="920" height="483" alt="Image" src="https://github.com/user-attachments/assets/d47fb31e-ca01-49b1-a816-da85a2b33fe0" />

1. Tok_embedding Weights check

<img width="1090" height="572" alt="Image" src="https://github.com/user-attachments/assets/e2decc0e-e516-4b2f-bc21-8199b03b46ac" />


2. Optimizer check

<img width="852" height="506" alt="Image" src="https://github.com/user-attachments/assets/81fbd2d6-3a23-4248-adf7-2d37e4db618c" />



## TP = 4 (TP only)

TP only didn’t reproduce this issue - So this issue might relate to FSDPModule

<img width="758" height="379" alt="Image" src="https://github.com/user-attachments/assets/1361692d-effb-420d-a922-438eca825111" />

**Comment by fegin** [comment]
Add some experiments.

1. I added following code after the checkpoint save in the trainer and set the saving frequency to 10, the loss curve is fine.
```
if self.step % 10 == 0:
   self.checkpointer.load(self.step)
```

2. A similar save and load frequency, but between save and load, I killed the trainer, then I can see loss jump a little for the first two steps right after load.

cc., @weifengpy, @wwwjn

**Comment by weifengpy** [comment]
> Hi [@wwwjn](https://github.com/wwwjn), thank you for your support! I’ve prepared a reproduction script based on the [latest main branch](https://github.com/pytorch/torchtitan/commit/a44dff1a41f6c0d8e504919ce4b1b50d05102f01), along with some instructions. Here is the [code](https://github.com/speed1313/torchtitan/commit/08fb43479eedc5016383cd4db628b9d38465a25d#diff-78cc79291e219fdfd73f7b1c7b5d442d1346f821b8add32e0e02d62597fe0ee5). I ran it with FSDP (degree = 8), and confirmed that the bug still exists. <img alt="Image" width="1307" height="702" src="https://private-user-images.githubusercontent.com/69028964/471775218-779e43b4-5227-4a2a-ab16-a050c9c367df.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU4ODIzNzcsIm5iZiI6MTc1NTg4MjA3NywicGF0aCI6Ii82OTAyODk2NC80NzE3NzUyMTgtNzc5ZTQzYjQtNTIyNy00YTJhLWFiMTYtYTA1MGM5YzM2N2RmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODIyVDE3MDExN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIxOWJhYWE0NzUzMWFkZDRhN2NmMDRlMGVhZGUzNjcwODQ3ZWUyZDY1MTkyYzJmYjZjMjQ3MGFjYTg5OWE5NDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fkakvBVubl2thlBzmZpEfQ1PO9uWjgbYVDEQI3TPHgU">

thanks for the repro! it helps a lot

**Comment by thuliu-yt16** [comment]
The problem still exists and in my experiment, I observed a similar loss increase after resuming and the significant loss decrease within several steps.

> I also encountered a loss discrepancy when resuming from checkpoint after setting the vocab size to be indivisible by the world size when I use FSDP. Since the loss quickly returned to the loss before resuming, I suspect there's a bug in how the optimizer state is saved.
> 
> <img alt="Image" width="475" height="383" src="https://private-user-images.githubusercontent.com/69028964/469572010-55dc6e7f-46e4-4993-925a-bcb1768f421a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0NTU5NTcsIm5iZiI6MTc1NjQ1NTY1NywicGF0aCI6Ii82OTAyODk2NC80Njk1NzIwMTAtNTVkYzZlN2YtNDZlNC00OTkzLTkyNWEtYmNiMTc2OGY0MjFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODI5VDA4MjA1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2NTEwMzkxMDg3NGRkMTJiOGI5OTk5YWJiYWI0OWMxYTBiNTkwZTQwYmE4ZTliYWVhODM4ZjE2YjI3NDAzYWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.rKJVMJLntoOj28M30e82c8i0gj7S14ZR7bGYHiU8qnE">
> By padding the vocabulary size to be divisible by the world size, the issue was resolved.
> 
> def pad_vocab_to_divisible(tokenizer, world_size: int):
>     current_vocab_size = len(tokenizer)
>     remainder = current_vocab_size % world_size
>     if remainder == 0:
>         return
>     
>     num_to_add = world_size - remainder 
>     current_reserved_special_token_num = sum(
>         1 for token in tokenizer.get_vocab().keys() if "reserved_special_token" in token
>     )
>     new_tokens = [
>         f"<|reserved_special_token_{current_reserved_special_token_num + i}|>"
>         for i in range(num_to_add)
>     ]
>     tokenizer.add_special_tokens({"additional_special_tokens": new_tokens})

**Comment by weifengpy** [comment]
@thuliu-yt16 @mingjihantencent @speed1313 @weixuansun disabling cached_state_dict resolves the problem: https://github.com/pytorch/torchtitan/pull/1702

this should unblock people for urgent needs. But I still need to dig deeper to understand how it's related to uneven sharding

**Comment by weifengpy** [comment]
the fix is landed to pytorch: https://github.com/pytorch/pytorch/pull/163130

@wwwjn do you need more to close the issue?

**Comment by wwwjn** [comment]
Update: Verified it works well! Thanks @weifengpy for help fixing the issue. 
<img width="1188" height="600" alt="Image" src="https://github.com/user-attachments/assets/9e505b33-c0b5-4352-88e6-3f4338de89c0" />

<img width="1188" height="600" alt="Image" src="https://github.com/user-attachments/assets/0c219c77-cbfd-47b4-8962-b5e8624f2050" />
